protos/graph.proto

```diff
@@ -0,0 +1,208 @@
+        // Node that executes when join nodes finishes executing
+        Node merged_node = 2;
+    }
+    */
+
+    /* Math */
+    // Arithmetic 
+    message Add {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Sub {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Mul {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Div {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Max {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Min {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Abs {
+        Node x = 1;
+    }
+    message Pow {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Mod {
+        Node x = 1;
+        Node y = 2;
+    }
+    // Trigonometry
+    message Sin {
+        Node x = 1;
+    }
+    message ArcSin {
+        Node x = 1;
+    }
+    message Cos {
+        Node x = 1;
+    }
+    message ArcCos {
+        Node x = 1;
+    }
+    message Tan {
+        Node x = 1;
+    }
+    message ArcTan {
+        Node x = 1;
+    }
+    // Random number generation
+    // returns a random float between low and high (inclusive)
+    message Random {
+        Node low = 1;
+        Node high = 2;
+    }
+
+    /* Boolean */
+    message And {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Or {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Not {
+        Node x = 1;
+    }
+    message Eq {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Gt {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Lt {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Ge {
+        Node x = 1;
+        Node y = 2;
+    }
+    message Le {
+        Node x = 1;
+        Node y = 2;
+    }
+    
+    // the operation that this node encapsulates
+    oneof op {
+        Const const_op = 1;
+        Retrieve retrieve_op = 2;
+        Mutate mutate_op = 3;
+        Switch switch_op = 4;
+        //Fork fork_op = 5;
+        //Merge merge_op = 6;
+        Add add_op = 7;
+        Sub sub_op = 8;
+        Mul mul_op = 9;
+        Div div_op = 10;
+        Max max_op = 11;
+        Min min_op = 12;
+        Abs abs_op = 13;
+        Pow pow_op = 14;
+        Mod mod_op = 15;
+        Sin sin_op = 16;
+        ArcSin arcsin_op = 17;
+        Cos cos_op = 18;
+        ArcCos arccos_op = 19;
+        Tan tan_op = 20;
+        ArcTan arctan_op = 21;
+        Random random_op = 22;
+        And and_op = 23;
+        Or or_op = 24;
+        Not not_op = 25;
+        Eq eq_op = 26;
+        Gt gt_op = 27;
+        Lt lt_op = 28;
+        Ge ge_op = 29;
+        Le le_op = 30;
+    }
+}
+
+// Defines a computation/evaluatable graph that takes in inputs from the component state
+// and when evaluated, mutates the component state.
+message Graph {
+    // List of input nodes - retrieved values from component state.
+    // Input nodes must all be reachable from the output nodes
+    repeated Node.Retrieve inputs = 1;
```

Comment:
After these retrieve nodes are evaluated, what happens? How is it linked to the mutate nodes?

Reply:
> After these retrieve nodes are evaluated, what happens?

Retrieve nodes retrieves a value from ECS's component state when evaluated. This value is then available for use for computation in the rest of the graph.

> How is it linked to the mutate nodes?


 If we think of retrieve nodes as the graph taking inputs by taking values from the component state and  mutate nodes as graph producing outputs by mutating the component state:
- Input nodes have to be connected to output nodes in the computation graph.
- It does not make sense to have an input retrieve node that is not connected and there by used to produce an output. (the input node is orphaned).

Reply:
How are input nodes connected to output nodes? How would an interpreter evaluate the graph?
I would imagine that a computation graph would only have mutate nodes. There is no point retrieving data and not using it to mutate data as you have said (input nodes must be connected to output nodes). So, why do we need to separate input nodes from output nodes?

Functional lenses do something similar:
```
set(get("position", "x"), 3);
set(get("position", "y") + get("speed", "y"), 3);
...
```

Reply:
> How are input nodes connected to output nodes? 

Input nodes are connected to output nodes by an inverse dependency of the output node on an input node.
> How would an interpreter evaluate the graph?

Populate the input nodes with values, recursively walk though the graph from the output nodes to the input nodes, start evaluating from the node with no dependencies.
> So, why do we need to separate input nodes from output nodes?

Technically, we don't really need this, some weak reasons for having it are:
- we know what the component dependencies of graph are.
- culling of redundant nodes will potentially already happen on the SDK side, so we will already know what the input nodes are.

I happy with removing it.

Reply:
I think we can leave it for now.